### PR TITLE
[Flash] Flash data scrambling

### PR DESCRIPTION
**Pull request by tjaychen**
This adds the prince based flash scrambling to the flash controller.
It has been **lightly** tested by making the 2nd bank scramble / de-scramble always.

The exact programmer's model of flash scrambling and the placement of the address filter is not yet decided, please see #2630 
Since OTP does not exist on master, the connection is also hard-wired instead of coming from flops.

For bronze specifically, we need to update the constraints to inform the tool that `prim_prince` is supposed to be multi-cycle, I haven't looked into that yet.

### CODE REVIEW

**Comment by tjaychen** [comment]
can you guys help review the second commit "for now" with an eye on bronze? 
Once we get it in for bronze, then we should probably fully review this. 

Of course, if you notice things we should fix / change for master, feel free to point it out also.

@sriyerg 
long term, this is probably something we need to heavily stress in block level.  Especially jumping back and forth between accesses for scrambled and non-scrambled. 

@msfschaffner 
I got over half way through until i realized the doc said we were going to use present.  Do you remember if that's what we were going to stick with? Or we were going to run it by a lab for feedback?

**Comment by eunchan** [comment]
Acked for constraints. we can add that into `hw/top_earlgrey/syn/constraints.sdc`

**Comment by tjaychen** [comment]
yeah i'm working on the fpga one right now.

**Comment by tjaychen** [comment]
okay fpga constraints updated. 
Let me know if you think this is good enough for bronze, i can move it over and modify the constraints.

File: hw/top_earlgrey/data/pins_nexysvideo.xdc [hunk 1]
```diff
@@ -11,6 +11,11 @@ create_generated_clock -name clk_50_unbuf -source [get_pin clkgen/pll/CLKIN1] [g
 create_generated_clock -name clk_48_unbuf -source [get_pin clkgen/pll/CLKIN1] [get_pin clkgen/pll/CLKOUT1]
 set_clock_groups -group clk_50_unbuf -group clk_48_unbuf -asynchronous
 
+## Preserve prim_prince modules and setup multi-cycle paths
+set_property KEEP_HIERARCHY TRUE [get_cells top_earlgrey/u_flash_eflash/gen_flash_banks[*].i_core/u_scramble/u_cipher]
+set_multicycle_path -setup 2 -through [get_pins -of_objects [get_cells top_earlgrey/u_flash_eflash/gen_flash_banks[*].i_core/u_scramble/u_cipher]]
```
[the remark this answers is not in the document]
Comment (cdgori) [reply hw/top_earlgrey/data/pins_nexysvideo.xdc:16]: it's been a long time, but my recollection is that usually you would need a `set_multicycle_path -hold 1` with the same args as your existing setup MCP, unless you still want the hold check on the zero cycle (which might be the case here, I haven't thought about it much).  (Basically my MCP files always had this n/n-1 pattern to them.)

File: hw/ip/flash_ctrl/rtl/flash_phy_rd.sv [hunk 2]
```diff
@@ -13,25 +13,26 @@
...
 //
-// TBD: Add support for descramble stage
-// The allocate and descramble indication received at read stage must be saved.
+// Support for descramble stage
+// The allocate and descramble indication received at read stage are saved.
 // When the read completes, depending on the 'descramble' indication saved, the
 // data is either stored into FIFO (reg + skid) between read and descramble stage,
 // or forwarded directly to the buffers (no de-scramble)
 //
 // If the storage element between read and de-scramble stages are completely fully
```
Comment (cdgori) [inline hw/ip/flash_ctrl/rtl/flash_phy_rd.sv:22]: fully -> full

File: hw/ip/flash_ctrl/rtl/flash_phy_rd.sv [hunk 3]
```diff
@@ -263,7 +290,111 @@ module flash_phy_rd import flash_phy_pkg::*; (
...
+  assign scrambled_data_o = fifo_data ^ mask;
+
+  // muxed data
+  assign muxed_data = hint_descram ? descrambled_data_i ^ mask : data_i;
+
+  // muxed data valid
+  // if no de-scramble required, return data on read complete
+  // if data is all empty (erased), also return data on read complete
+  // if descramble is required, return data when descrambler finishes
+  assign data_valid = forward | fifo_data_ready;
```
Comment (cdgori) [inline hw/ip/flash_ctrl/rtl/flash_phy_rd.sv:397]: `data_valid` seems a little too generic?  I originally thought maybe `rd_data_valid` but I'm not convinced of that one either here.  Have a think on it.

File: hw/top_earlgrey/data/pins_nexysvideo.xdc (hunk 1 above)
[the remark this answers is not in the document]
Comment (tjaychen) [reply hw/top_earlgrey/data/pins_nexysvideo.xdc:16]: you are right :)
I put the comment at a bad place.  If you scroll past our comments you will see the hold `hold` case.

[the remark this answers is not in the document]
Comment (cdgori) [reply hw/top_earlgrey/data/pins_nexysvideo.xdc:16]: Hah - the way this wrapped on my screen I didn't see the next line.  I guess I was correct. :)

File: hw/ip/flash_ctrl/rtl/flash_phy_rd.sv (hunk 3 above)
Comment (tjaychen) [reply hw/ip/flash_ctrl/rtl/flash_phy_rd.sv:397]: sounds good. Yeah i didn't attach a 'rd' prefix because this is already the read module.
I at one point tried to make it specific, then it turned into this gigantic string. I'll think on it.

**Comment by sriyerg** [comment]
> can you guys help review the second commit "for now" with an eye on bronze?
> Once we get it in for bronze, then we should probably fully review this.
> 
> Of course, if you notice things we should fix / change for master, feel free to point it out also.
> 
> @sriyerg
> long term, this is probably something we need to heavily stress in block level. Especially jumping back and forth between accesses for scrambled and non-scrambled. At the top level I'm not really able to create that kind of traffic, since the controller is sequential and ibex is too slow right now (until we turn icache one). I'll find some time to sync up with you on how I might be able to leverage the block level bench.
> 
> @msfschaffner
> I got over half way through until i realized the doc said we were going to use `present`. Do you remember if that's what we were going to stick with? Or we were going to run it by a lab for feedback?
> 
> @eunchan
> see my point about `prim_prince` constraints above, I'll check-in with you on this. Let me see if I can figure out the right constraints for FPGA first.

Acknowledged.

File: hw/top_earlgrey/data/pins_nexysvideo.xdc (hunk 1 above)
[the remark this answers is not in the document]
Comment (cdgori) [reply hw/top_earlgrey/data/pins_nexysvideo.xdc:16]: > @cdgori
> @msfschaffner
> @eunchan
> 
> does this look right to you guys? the `prim_prince` module is completely combo, so i figured making it multicycle path through all pins should be okay.

Just double checking, all of the inputs are stable for the 2 clocks (key_i/data_i/dec_i)?  No chance that dec_i is late arriving or something like that?

File: hw/ip/flash_ctrl/rtl/flash_phy_rd.sv (hunk 3 above)
Comment (cdgori) [reply hw/ip/flash_ctrl/rtl/flash_phy_rd.sv:397]: that's actually a good argument for leaving it as-is, I hadn't considered that.  I only noticed it a couple lines down when you were swapping out `rd_done` for `data_valid` but since it's the read pipeline it's kind of implied.

Comment (tjaychen) [reply hw/ip/flash_ctrl/rtl/flash_phy_rd.sv:397]: yeah the inputs `data` and `dec` are coming from fifo'd outputs and not popped until descrambling. 
`key` is hardwired at the moment..but it should long term come from otp flops that are static after initial sense. 

Honestly...i don't like relying on multi-cycle path to make things work... there's just no clean way to ensure one changes when the other changes... i'd rather have things to be correct by construction.. and you can see I have a giant comment regarding that.

File: hw/top_earlgrey/data/pins_nexysvideo.xdc (hunk 1 above)
[the remark this answers is not in the document]
Comment (tjaychen) [reply hw/top_earlgrey/data/pins_nexysvideo.xdc:16]: yeah the inputs data and dec are coming from fifo'd outputs and not popped until descrambling.
key is hardwired at the moment..but it should long term come from otp flops that are static after initial sense.

Honestly...i don't like relying on multi-cycle path to make things work... there's just no clean way to ensure design changes propagate to constraints... i'd rather have things be correct by construction.. and you can see I have a giant comment regarding that.

[the remark this answers is not in the document]
Comment (cdgori) [reply hw/top_earlgrey/data/pins_nexysvideo.xdc:16]: > you are right :)
> I put the comment at a bad place. If you scroll past our comments you will see the hold `hold` case.

I removed my comment, since I did eventually scroll down and see you had done it already :)

[the remark this answers is not in the document]
Comment (cdgori) [reply hw/top_earlgrey/data/pins_nexysvideo.xdc:16]: > yeah the inputs data and dec are coming from fifo'd outputs and not popped until descrambling completes.
> key is hardwired at the moment..but it should long term come from otp flops that are static after initial sense.
> 
> Honestly...i don't like relying on multi-cycle path to make things work... there's just no clean way to ensure design changes propagate to constraints... i'd rather have things be correct by construction.. and you can see I have a giant comment regarding that.

I hate them and try to remove them everywhere I can (sometimes you just can't but I try).  They can easily get you a dead chip because nothing is keeping constraints in sync with design as you say, and nothing is simulating constraints.

[the remark this answers is not in the document]
Comment (tjaychen) [reply hw/top_earlgrey/data/pins_nexysvideo.xdc:16]: @msfschaffner .. say.... should we consider pipelining inside `prim_present` and `prim_prince` directly? :)
i ended up going this approach for `gf_mult`... it is more expensive to do (since you need storage inside).. and there are limits to how you can carve it (should probably be divisible).. but you know..no need for constraint tweaking :)

I have to admit though, if not for constraints, the multi-cycle approach in this case is pretty appealing, since it would also solve another problem i had to work-around.

**Comment by msfschaffner** [comment]
> 
> @msfschaffner
> I got over half way through until i realized the doc said we were going to use `present`. Do you remember if that's what we were going to stick with? Or we were going to run it by a lab for feedback?
> 

sorry I missed this one. I think we settled on PRINCE since we can potentially use the full-round version for flash scrambling...

File: hw/top_earlgrey/data/pins_nexysvideo.xdc (hunk 1 above)
[the remark this answers is not in the document]
Comment (msfschaffner) [reply hw/top_earlgrey/data/pins_nexysvideo.xdc:16]: actually, I've been thinking about placing a register halfway into `prim_prince` in order to optimize timing in #2655. would that work for you as well? since the key is static, we probably do not have to pipeline the key into the second cycle in `prim_prince`, hence you only need 64 flops per `prim_prince` instance.

File: hw/ip/flash_ctrl/rtl/flash_ctrl_pkg.sv [hunk 4]
```diff
@@ -110,4 +116,22 @@ package flash_ctrl_pkg;
...
+  // otp to flash_phy
+  typedef struct packed {
+    logic [127:0] addr_key;
+    logic [127:0] data_key;
+  } otp_flash_t;
+
+  // default value of otp_flash_t
+  parameter otp_flash_t OTP_FLASH_DEFAULT = '{
+    addr_key: 128'hDEADBEEFBEEFFACEDEADBEEF5A5AA5A5,
+    data_key: 128'hDEADBEEF5A5AA5A5DEADBEEFBEEFFACE
```
Comment (msfschaffner) [inline hw/ip/flash_ctrl/rtl/flash_ctrl_pkg.sv:132]: I like the `DEADBEEF_BEAFFACE` ;)

File: hw/top_earlgrey/data/pins_nexysvideo.xdc (hunk 1 above)
[the remark this answers is not in the document]
Comment (tjaychen) [reply hw/top_earlgrey/data/pins_nexysvideo.xdc:16]: that would definitely work for me :) Are you thinking about placing a flop at the reflect point?

File: hw/ip/flash_ctrl/rtl/flash_phy_rd.sv [hunk 5]
```diff
@@ -263,7 +290,111 @@ module flash_phy_rd import flash_phy_pkg::*; (
...
+  logic hint_descram;
+  logic [NumBuf-1:0] alloc_q2;
+
+  assign scramble_stage_rdy = data_fifo_rdy & mask_fifo_rdy;
+
+  // data is consumed when:
+  // 1. When descrambling completes
+  // 2. Immediately consumed when descrambling not required
+  // 3. In both cases, when data has not already been forwarded
+  assign fifo_data_ready = hint_descram ? descramble_req_o & descramble_ack_i & !hint_forward :
```
Comment (msfschaffner) [inline hw/ip/flash_ctrl/rtl/flash_phy_rd.sv:311]: I know this is nitty :), but I think this should either follow the  `&& !` or `& ~` convention

File: hw/ip/flash_ctrl/rtl/flash_phy_rd.sv [hunk 6]
```diff
@@ -263,7 +290,111 @@ module flash_phy_rd import flash_phy_pkg::*; (
...
+  // is that the mask / data are always pushed, it is then selectively popped based
+  // on the forward / de-scrambling hints.
+  //
+  // All these problems could be resolved if the timings matched exactly, however
+  // the user would need to correctly setup constraints on either flash / gf_mult
+  // timing change.
+  prim_fifo_sync #(
+      .Width  (DataWidth + 2 + NumBuf),
+      .Pass   (0),
+      .Depth  (2)
```
Comment (msfschaffner) [inline hw/ip/flash_ctrl/rtl/flash_phy_rd.sv:337]: nit: indentation seems to be off...

File: hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv [hunk 7]
```diff
@@ -0,0 +1,91 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Flash Phy Scramble Module
+//
+// This module implements the flash scramble / de-scramble operation
+// This operation is actually XEX.  However the comoponents are broken
```
Comment (msfschaffner) [inline hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv:8]: com~~o~~ponents

File: hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  input [DataWidth-1:0] plain_data_i,
+  input [DataWidth-1:0] scrambled_data_i,
+  input [KeySize-1:0] addr_key_i,
+  input [KeySize-1:0] data_key_i,
+  output logic calc_ack_o,
+  output logic op_ack_o,
+  output logic [DataWidth-1:0] mask_o,
+  output logic [DataWidth-1:0] plain_data_o,
+  output logic [DataWidth-1:0] scrambled_data_o
+  );
```
Comment (msfschaffner) [inline hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv:28]: nit: unneeded spaces

File: hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    if (!rst_ni) begin
+      cnt <= '0;
+    end else if (op_req_i && op_ack_o) begin
+      cnt <= '0;
+    end else if (op_req_i) begin
+      cnt <= cnt + 1'b1;
+    end
+  end
+
+  // Previous discussion settled on PRESENT, using PRINCE here for now
```
Comment (msfschaffner) [inline hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv:70]: I think we went for full-round PRINCE in the end. If you use PRINCE, I suggest to use the new keyschedule to prevent attacks described in https://eprint.iacr.org/2014/656.pdf: `.UseOldKeySched(1'b0)`...

File: hw/ip/flash_ctrl/rtl/flash_phy_rd.sv (hunk 5 above)
Comment (tjaychen) [reply hw/ip/flash_ctrl/rtl/flash_phy_rd.sv:311]: o is this because the ternary statement is technically a conditional?

Comment (tjaychen) [reply hw/ip/flash_ctrl/rtl/flash_phy_rd.sv:311]: o wait, you just meant that i should be consistent. Got it.

Comment (msfschaffner) [reply hw/ip/flash_ctrl/rtl/flash_phy_rd.sv:311]: yes I do not mind much which is used :)

**Comment by msfschaffner** [comment]
FYI: I've added a parameter to `prim_prince` to optionally instantiate a pipelining register halfway in the data path, see https://github.com/lowRISC/opentitan/pull/2678.

**Comment by tjaychen** [comment]
i'll probably keep this PR as is for now, and add that in (as well as the
key schedule thing) as a separate PR.

**Comment by msfschaffner** [comment]
ok, sounds good

File: hw/top_earlgrey/data/pins_nexysvideo.xdc (hunk 1 above)
[the remark this answers is not in the document]
Comment (msfschaffner) [reply hw/top_earlgrey/data/pins_nexysvideo.xdc:16]: yes exactly - see https://github.com/lowRISC/opentitan/pull/2678.

File: sw/device/lib/flash_ctrl.c [hunk 10]
```diff
@@ -131,6 +131,10 @@ int flash_read(uint32_t addr, part_type_t part, uint32_t size, uint32_t *data) {
   return get_clr_err();
 }
 
+void flash_cfg_scramble_en(bool en) {
+  REG32(FLASH_CTRL_SCRAMBLE_EN(0)) = en ? 1 : 0;
```
Comment (mcy) [inline sw/device/lib/flash_ctrl.c:135]: Consider using the modern style with `mmio_region_t`.

Also, the ternary is hard to parse. Use something like `en & 1` instead.

File: sw/device/lib/flash_ctrl.h [hunk 11]
```diff
@@ -96,6 +96,11 @@ int flash_read(uint32_t addr, part_type_t part, uint32_t size, uint32_t *data);
  */
 void flash_cfg_bank_erase(bank_index_t bank, bool erase_en);
 
+/**
+ * Configure scramble enable
+ */
+void flash_cfg_scramble_en(bool en);
```
Comment (mcy) [inline sw/device/lib/flash_ctrl.h:102]: `en` -> `enable`.

File: sw/device/lib/flash_ctrl.c (hunk 10 above)
Comment (lenary) [reply sw/device/lib/flash_ctrl.c:135]: There's no DIF for flash control, so I suggest leaving the software as-is in this patch, and we can revisit it once we have a DIF.

Comment (cdgori) [reply sw/device/lib/flash_ctrl.c:135]: > Also, the ternary is hard to parse. Use something like `en & 1` instead.

Not a crucial point, but I think this is definitely a "hardware vs software engineer" thing - I find that `en & 1` much harder to parse :) - but then I am accustomed to a world of multi-nested ternaries...

Comment (tjaychen) [reply sw/device/lib/flash_ctrl.c:135]: @mcy, to @lenary 's point, i was mostly just trying to maintain the current style until a DIF is officially created.
I'll address your other comments though.

File: hw/ip/flash_ctrl/rtl/flash_phy_rd.sv (hunk 2 above)
Comment (tjaychen) [reply hw/ip/flash_ctrl/rtl/flash_phy_rd.sv:22]: fixed.

File: hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv (hunk 9 above)
Comment (tjaychen) [reply hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv:70]: @msfschaffner is this referring to appendix C of the paper?
Because it looks like when i use the new schedule, the decrypt results and the encrypt results are not the same.

File: sw/device/lib/flash_ctrl.h (hunk 11 above)
Comment (tjaychen) [reply sw/device/lib/flash_ctrl.h:102]: sorry i can't tell if this is referring to the function name or the variable. 
I'm guessing function name?

File: hw/ip/flash_ctrl/rtl/flash_phy_rd.sv (hunk 5 above)
Comment (tjaychen) [reply hw/ip/flash_ctrl/rtl/flash_phy_rd.sv]: fixed

File: hw/ip/flash_ctrl/rtl/flash_phy_rd.sv (hunk 6 above)
Comment (tjaychen) [reply hw/ip/flash_ctrl/rtl/flash_phy_rd.sv]: fixed

File: hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv (hunk 7 above)
Comment (tjaychen) [reply hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv]: fixed

File: hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv (hunk 8 above)
Comment (tjaychen) [reply hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv]: fixed

File: sw/device/lib/flash_ctrl.h (hunk 11 above)
Comment (mcy) [reply sw/device/lib/flash_ctrl.h]: As Sam points out, this doesn't Actually Matter. Up to you. =)

File: hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv (hunk 9 above)
Comment (msfschaffner) [reply hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv:57]: yes exactly. although I may have done something wrong there. let me double check.

if `.UseOldKeySched(1'b1)` is working, can we just stick in a TODO that mentions that this should be changed to `.UseOldKeySched(1'b0)` once that mode is working?

Comment (tjaychen) [reply hw/ip/flash_ctrl/rtl/flash_phy_scramble.sv:57]: so when i looked at it, it seemed the round key was supposed to be alternating, thus creating a symmetric structure.
I didn't get the feeling when looking at this one, but I didn't read very closely.

I'll stick with old key schedule for now, and switch later.
